### ISSUE TITLE
HAProxy swaps headers between connections when using HTX

### ISSUE THREAD

**Issue by veggiespam** [haproxy/haproxy#290]
## Output of `haproxy -vv` and `uname -a`

```
HA-Proxy version 2.0.4 2019/08/06 - https://haproxy.org/
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declara
tion -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-d
ereference
  OPTIONS = USE_PCRE2_JIT=1 USE_STATIC_PCRE2=1 USE_OPENSSL=1 USE_SLZ=1 USE_SYSTEMD=1

Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER -PCRE -PCRE_JIT -PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED -REGPARM -STATIC
_PCRE +STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL +GETADDRINFO +OPENSSL -LUA +FUTEX +ACCEPT4 -MY_ACCEPT4 -ZLIB +SLZ +CPU_AF
FINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=2).
Built with OpenSSL version : OpenSSL 1.1.1c  28 May 2019
Running on OpenSSL version : OpenSSL 1.1.1c  28 May 2019
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with network namespace support.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with PCRE2 version : 10.31 2018-02-12
PCRE2 library su

Linux macaw-1d 4.20.8-1.el7.elrepo.x86_64 #1 SMP Tue Feb 12 18:03:03 EST 2019 x86_64 x86_64 x86_64 GNU/Linux
```

## What's the configuration?

```
global
    maxconn             10240
    daemon
    nbproc              1
    tune.ssl.default-dh-param 2048
    ssl-mode-async
    stats socket /var/run/haproxy.sock group nobody mode 770 expose-fd listeners
    log 127.0.0.1       local0
    log 127.0.0.1       local1 err
    hard-stop-after     10m
    user nobody
    group nobody
    chroot /var/empty

resolvers internal
        nameserver resolver1 REDACTED
        nameserver resolver2 REDACTED
        nameserver resolver3 REDACTED
    
defaults
    default-server init-addr last,libc,none
    log     global
    mode    http
    option  httplog
    option  dontlognull
    option  forwardfor
    option  http-keep-alive
    option  redispatch
    log-format [%t]\ transport=%ft\ client_ip=%ci\ backend_ip=%bi\ backend_port=%bp\ backend_pool=%b\ server_name=%s\ status_code=%ST\ request=%{+Q}r\ time_waiting_ms=%Tw\ time_to_established_ms=%Tc\ time_session_ms=%Tt\ bytes_read=%B\ term_state=%tsc\ ssl_ciphers=%sslc\ ssl_version=%sslv\ x_unique_id=%ID\ captured_request_duid=%{+Q}[capture.req.hdr(0)]\ captured_response_duid=%{+Q}[capture.res.hdr(0)]
    retries 3
    stats   enable
    timeout client 60s
    timeout http-request 10s
    timeout connect 4s
    timeout server 60s
    unique-id-format %{+X}o\ %ci:%cp_%fi:%fp_%Ts_%rt:%pid

frontend sftp
    bind 0.0.0.0:1234
    mode tcp
    no option dontlognull
    log-format [%t]\ transport=%ft\ client_ip=%ci\ backend_ip=%bi\ backend_port=%bp\ backend_pool=%b\ server_name=%s\ time_waiting_ms=%Tw\ time_to_established_ms=%Tc\ time_session_ms=%Tt\ bytes_read=%B\ bytes_uploaded=%U\ term_state=%ts
    default_backend sftp

backend sftp
    mode tcp
    option tcpka
    balance leastconn
    server task_0 REDACTED:39889 check port 39889 send-proxy

frontend stats
    bind 0.0.0.0:2345
    maxconn 128
    option http-use-htx
    option httpclose
    http-request use-service prometheus-exporter if { path /metrics }
    stats refresh 5s
    stats uri /stats
    stats show-node
    stats realm HAProxy\ Statistics

frontend http-in
    bind *:80
    monitor-uri /site_alive
    acl php_request path_end -i .php
    http-request deny if php_request

    http-request redirect code 301 scheme https

frontend https-in
    mode http
    bind *:443 ssl no-sslv3 crt /etc/pki/tls/certs/wildcard.pem ciphers CDHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-SHA384:ECDHE-RSA-AES128-SHA256:ECDHE-RSA-AES256-SHA:ECDHE-RSA-AES128-SHA:DHE-RSA-AES256-SHA256:DHE-RSA-AES128-SHA256:DHE-RSA-AES256-SHA:DHE-RSA-AES128-SHA:AES256-GCM-SHA384:AES128-GCM-SHA256:AES256-SHA256:AES128-SHA256:AES256-SHA:AES128-SHA:HIGH:!aNULL:!eNULL:!EXPORT:!DES:!MD5:!PSK:!RC4:!3DES alpn h2,http/1.1

    http-request set-header X-Unique-ID %[unique-id]

    declare capture response len 42
    declare capture request len 42
    http-request capture req.cook(duid),sha1,hex id 0
    http-response capture res.cook(duid),sha1,hex id 0

    acl host_REDACTED hdr(host) -i REDACTED.REDACTED
    acl xframe_options res.hdr(X-Frame-Options) -m found
    http-response set-header Strict-Transport-Security max-age=31536000
    http-response set-header X-Content-Type-Options nosniff
    http-response set-header X-Frame-Options SAMEORIGIN if !host_REDACTED !xframe_options

    acl php_request path_end -i .php
    http-request deny if php_request

    acl vpn_ips src REDACTED
    acl office_and_vpn_allowed src REDACTED

    acl host_mobile-api_REDACTED hdr(host) -i mobile-api.REDACTED
    use_backend mobile if host_mobile-api_REDACTED

backend mobile
    redirect scheme https if !{ ssl_fc }
    balance roundrobin
    server task_0 REDACTED:31398 ssl check
    server task_1 REDACTED:31183 ssl check
    server task_2 REDACTED:31589 ssl check
    server task_3 REDACTED:37684 ssl check
    server task_4 REDACTED:34016 ssl check
    server task_5 REDACTED:30385 ssl check
```

## Steps to reproduce the behavior

1. Set up haproxy with systemd
2. Template routing configuration to backends based on host:port determined via service discovery (frequent updates) with hitless reload for updates
3. Generate traffic of approximately 50 requests per second
4. Observe headers (such as Set-Cookie) being sent to the incorrect requestor

## Actual behavior

Similar to #116, we noticed headers being returned to the wrong clients. Note the config pasted above is a snippet, we have around 100 backends that are conditionally used by same frontend.

Our analysis showed under minimal load in a master/worker configuration of HAProxy, headers would get swapped.  We did not use tcpdump per suggestions in #116, but used code in the middleware of our web application which detected these header swaps.  We also observed that the swaps usually occurred in groups, for example, 10 swaps might occur in a 1-minute interval during load and will otherwise be normal.  

The header swaps began when we upgraded from version 1.8.9, which does not support HTX, to 2.0.4, which enables it by default.  There were no other changes to our infrastructure.

## Expected behavior

Request/response contexts to be handled correctly.

## Do you have any idea what may have caused this?

Something in HTX processing per below.

## Do you have an idea how to solve the issue?

Disabling HTX via global `no option http-use-htx` has resolved this issue in our configuration. 

We are concerned that in master/v2.0.5-dev, the option to disable HTX has been completely [removed](https://github.com/haproxy/haproxy/blob/03627245c67216bb003912df42150ec1401cd716/src/cfgparse-listen.c#L2010-L2019). We feel that there is a bug in HTX processing that only arises under certain load, potentially caused by frequent configuration reloads. The majority of all backends used here are HTTP 1.1 only if that helps.

**Comment by lukastribus** [comment]
2.0.6 is the latest stable release, you need to report issues against that, there are *TONS* of bugfixes after 2.0.4.

> We are concerned that in master/v2.0.5-dev, the option to disable HTX has been completely removed.

That's master (2.1-dev) only. You will be able to disable HTX in all current and future 2.0.X releases.

**Comment by veggiespam** [comment]
Thank you for the clarification about `no option http-use-htx` in v2.0.x.

We cannot upgrade to 2.0.6 due to issues described in #278.  When 2.0.7 is released, my team will try a deploy again.  There is different pointer arithmetic in the newer HTX code, so I appreciate the patches.  

Per our issues and #116, I recommend you post a security alert that HTX should be disabled in multiprocess/multithread implementations of HAProxy running version  <= 2.0.5 due to possible header swap.

**Comment by capflam** [comment]
That's interesting because I'm tracking a similar bug but with h2-to-h2 connections. Did you observed the bug on h1 client connections ? And could you explain how you observed it? In HAProxy logs only or on the client side too ? 

How many threads do you have ? And if possible, could you try to start HAProxy with only 1 thread ? It could be useful to know if it is a multithreaded issue or not. Finally, you could also try disabling connections pooling with the option `http-reuse never` and/or `max-session-srv-conns 0` in your frontend section.

**Comment by veggiespam** [comment]
We observed the situation by modifying our downstream system.  We basically inserted a device ID finger print (eg, User-Agent+stuff) into the browser data stream (form value or new header) and re-calculated this value with every request.  If the User-Agent for a request was different from the saved fingerprint, then something happened after which the connection and session would be terminated.  We can definitely see the problematic change-over in our detailed logs.

We discussed disabling multithreadedness, but the cost would be significant in the number of servers we'd have to spin up to accommodate our load.  We have attempted to replicated in a test environment without success.  Testing with `http-reuse never` or `max-session-srv-conns 0` is a non-option at this time since it would be playing with fire in production.

We've had `no option http-use-htx` in place for one week now and there has been no triggers whereas previously, there were many per day. With #292 also reporting a mix-up in data returned to end user, there is definitely a Confidentiality and Integrity problem somewhere in the system.  

Consider a security alert please.

**Comment by capflam** [comment]
I pushed a patch that may fix this bug. It was pushed to 2.0 and 1.9. @veggiespam, could you make some tests with this patch ?

**Comment by veggiespam** [comment]
Unfortunately, we are unable to deploy a "patch", we need a full release for software to make it into production.  When 2.0.7 becomes official, my team will attempt to deploy and I'll provide feedback.

Your patch definitely fixed some bug in https://github.com/haproxy/haproxy/commit/6884aa3eb00d1a5eb6f9c81a3a00288c13652938.  You can state in the release note that it "probably fixes the three issues" and update the release notes after confirmation from the each of the three teams æffected.

**Comment by capflam** [comment]
FYI, the release 2.0.7 was published.

**Comment by veggiespam** [comment]
Deployment plan:
1. Burn-in test in Dev/Test/Stage
1. Deploy in prod with HTX *disabled*
1. Let run for a bit
1. Enable HTX on single node
1. Observe to see if our triggers are fired
1. Slowly enable HTX on more nodes

Thanks for the 2.0.7 drop.

**Comment by veggiespam** [comment]
Update: We are currently on Step 3 above with v2.0.7 rolled out to all nodes.  We will enable Step 4 beginning Oct 7.  I will provide additional feedback around Oct 10.

**Comment by veggiespam** [comment]
After analyzing the results over the last 14 days, we found that we still see header swap in both 2.0.4 and 2.0.7 - this is while HTX was ***disabled*** in both cases.  However, we are receiving less swaps that we received in 2.0.4 when HTX was enabled.

We now feel that HTX is exacerbating another bug that exists somewhere in the software.  I would like guidance on next steps.

**Comment by capflam** [comment]
The fixed bug in the H2 multiplexer is not related to the HTX. It affects versions below the 2.0.7. I'm a bit surprised you still experience it on the 2.0.7. So I need more information. You said you are using a majority of HTTP/1.1 servers. On the server side, does the headers mix up happen for H2 connections only, H1 connections only, or both ?  Same question on the client side. The best of course would be to have network captures before and after HAProxy when the bug happens.

**Comment by veggiespam** [comment]
With further analysis, we discovered an edge case that we missed and eliminated most of the candidate header swaps.  However, we have a few instances where is still occurs not part of this edge case.  We are still investigating.  

HTX is enabled for some systems.

**Comment by wtarreau** [comment]
Well, given that there are two totally independent code paths for HTX and non-HTX not even reusing code, from the file descriptor to the analysers, at this point if you manage to find the same bug on both implementations, I would be very tempted to think that the bug is outside. I mean at this point, the probability that the same bug affects both implementations is the same as if you were testing with haproxy and any other proxy like nginx or varnish.

I would strongly suggest to switch to latest 2.1-dev and turn on traces in H1 or H2 to see exactly what's received/sent by haproxy. Alternately you could add ssl_bc_session_key/ssl_fc_session_key in your logs and decrypt the traffic with wireshark to see what's sent into haproxy and what goes out of it (there's an example here: https://gist.github.com/phemmer/31e89d1a3823230de1e1c47386ff508b).

**Comment by peterfroehlich** [comment]
We sadly had similar problems with 2.0.8. 

After a few trials on low volume LBs we deployed the release on our prod env. After that, reports started to come in that delivered content was getting mixed up (e.g. a svg was delivered instead of a requested js asset). We got around 200 reported cases out of ~40 million requests in a 24h time period. I guess the estimated number of unreported cases is much higher.

We are using linkerd1 in the backend and use a header fields to route requests to services. So swapping header fields is fatal. After we rolled back to 1.9.12 the problems vanished. We did not try 2.0.8 without HTX. H2 is used on the edge, H11 on the inside.

Sadly we found no way to force the error to happen and errors get only reported from the js logging, as the infra thinks the traffic is valid. We are implementing a way to log mismatched requests/responses on the client side to get better visibility.

**Comment by git001** [comment]
@peterfroehlich just for my curiosity, why is the order of headers important as long as all headers are available from client to server?

**Comment by capflam** [comment]
Thanks for the report. If I understand, the mix-up happens on the request headers and this leads to serve the wrong resource to the client. It is interesting because at this stage, I don't see how we can skip the decoding of a header frame without killing the connection itself. Till now, for me, this was the only way to have a headers mix-up.  I will review all the code again.

**Comment by capflam** [comment]
Hum, just saw @git001's comment. The mix-up happens between headers of different requests not between headers of a single request, right ? If it is a problem of order between the headers into a request, it is not related to this issue. Could you confirm please ?

**Comment by haproxy-mirror** [comment]
Also it would be very useful to know if this still happens in 2.0 with
"no option http-use-htx", and if it happens in 1.9 with the option set
so that we can bisect whether it sticks to the version or to a subset
of internal code.

**Comment by peterfroehlich** [comment]
> Hum, just saw @git001's comment. The mix-up happens between headers of different requests not between headers of a single request, right ? If it is a problem of order between the headers into a request, it is not related to this issue. Could you confirm please ?

The problem is related to a mixup between the headers of different requests, yes. We are using a header to route the request to different services. Swapping the headers of requests lead to a  wrongly routed request.  

> Also it would be very useful to know if this still happens in 2.0 with
"no option http-use-htx", and if it happens in 1.9 with the option set
so that we can bisect whether it sticks to the version or to a subset
of internal code.

I agree, this would be useful. We are using very aggressive caching headers for assets, so I'm reluctant to tinker with prod traffic and we weren't able to recreate the error in a lab env...
When we have implemented a reliable way to detect the problem without the users need to report it to us we can try again.

**Comment by wtarreau** [comment]
Guys, I'm having a question. Are you certain it cannot be on the server side that the application is mixing multiple requests from multiple clients over the same connection ? In 2.0 connection reused is enabled by default as it seems that the old problems of home-made bogus servers from the late 90s are long gone and all servers in field are now safe regarding connection reuse.

I'm asking because when the client connects to haproxy over H2, all requests are eligible to backend reuse, which is less the case in H1 (only the second and subsequent requests are, and only once the backend connection has proven support for keep-alive). And connection sharing is a bit less constrained in HTX than in legacy thus it can theorically happen slightly more often.

I'm asking because the way the code is architected makes it extremely unlikely that headers could be shared between different connections: there's no such entity as a header in the whole code, that could even accidently be reused. And when the client is using H2, we don't even use a headers table, the compression is entirely static so that even a bug in the compressor or missing frames or whatever cannot result in that effect. That's why I'm really starting to suspect that the unexpected headers are being passed along the whole chain and that we're looking for the problem at the wrong place.

**Comment by peterfroehlich** [comment]
I've looked into this, we are using 'option http-server-close' nearly everywhere so connection reuse to the server (akka-http in our case) should be disabled as I understand it...?

**Comment by wtarreau** [comment]
It should indeed. We'll really need exploitable traces to make any
progress on this, because there are still way too many variations between
situations where this is reported to happen to draw any conclusion, let
alone figure what to search for :-/

**Comment by peterfroehlich** [comment]
Just to update this, we are carefully testing config permutations with a small fraction of live traffic at the moment. 
Interestingly, it seems that only the lb clusters at the edge affect the problem, which leads to the conclusion that its related to HTTP2 on HTX (we are using HTTP11 inside). 
We will run 2.0.8 without HTX on the edge over the weekend, will keep you posted.

**Comment by capflam** [comment]
@peterfroehlich, thanks for these infos. If there is a bug, it should be in the HEADERS frame decoding and most probably in the HPACK code. But, by reading the code, nothing shocked me. And without any reproducer, it is just a hell to figure out what is happening.

So it is a good news if you can reproduce it. It could be really helpful to have a network capture that exhibits the bug. For now, I don't see any other solution to understand the problem. Have you a way to be sure the mixup happens for requests of a same client on one connection ? If it happens on the client side, it should be because the hpack header table is not shared between h2 connection. If not, we will have to rethink the issue from scratch.

**Comment by peterfroehlich** [comment]
We updated to 2.0.9, sadly no change: With HTX on, the error is present. With HTX off, its not. 

@capflam We have nearly 50M Req per day over the systems and our alert triggers about 20 times a day. We can't record everything that goes over the wire and need to find a way to pull traces if the problem occurs... 

@veggiespam Are you still having problems?

**Comment by veggiespam** [comment]
We are upgrading to 2.0.9 today but will keep HTX disabled for the moment.  There have been no header swaps in four weeks with HTX disabled using version 2.0.7.

**Comment by wtarreau** [comment]
OK so it seems to converge to H2+HTX causing the issue on request headers then (and yes, there we have HPACK). As to why it would affect only HTX at first thought it doesn't make sense, but it's not impossible that one tiny difference in a buffer size causes a wrapping in one case and not the other for example. Still extremely strange.

**Comment by veggiespam** [comment]
Upgrade complete and system running acceptably.  HTX is off and no swaps.

**Comment by git001** [comment]
When you enable HTX are then header swaps?

**Comment by peterfroehlich** [comment]
I just dissected a trace and found some strange behaviour that maybe links the problem to the new retry logic…? 
Some basics: 
- We are looking at the inside traffic from edge lb to varnish over HTTP11.
- Our JS based logging code in the client sends an UUID with each request and expects the same UUID in the response.
- Vanish has its caching disabled for the packets with UUID
- We are using “option http-server-close”

I got alerted of a swapped header: I found two GETS from a client, fetching the same page inside of 16ms inside one TCP stream. 
The first only registered in the edge logs with termination state “SD”
The second got forwarded to the infrastructure, produced multiple log entires but finished with termination state “CL”. 
The tcpdump shows the connection finishing with RST, ACK send from varnish to edge lb (Connection: close?).
The trace triggered because the frontend Code send the UUID of request one, but got the UUID from request two back.  
And as I understand “option http-server-close” should prevent that two requests are in the same tcp stream…? 
Thats why I think that maybe the new retry logic is to blame as “conn-failure” is the new default as documented. 

As a counter point that all this is maybe just a wierd connection issue triggering some bug in the JS: The client was a mobile device, so 16ms seems very fast for a retry. BUT, the requests contain different UUIDs, and these only get set by the JS in the browser. Does not explain why both requests are in the same tcp stream though.

I will try running with HTX on and retry-on completely disabled.

**Comment by capflam** [comment]
The retry on a connection failure is not a new behaviour. It exists since at least 10 years (maybe since the beginning). It is just mentioned as the default behaviour because of the new directive `retry-on` which allow you to enable other kind of retries.

To be sure I properly understand :

* You have an H2 client sending 2 requests on the same connection, with a UUID for each, let's say `uuid1` and `uuid2`. Both requests are sent in HTTP/1.1 to a varnish server, each one on a different connection because there is no keep-alive. 
* The first one is closed with the termination state "SD". So, it means the response headers have been received and the connection has been closed during the payload transfer. 
* The second request is closed with the termination state "CL". This one is pretty strange. Is there any response sent by varnish for this one or not ? Maybe when an error arises in your JS client, because of UUIDs mixup or anything else, it aborts the connection. It may explain this termination state.
* On the client side, `uuid2` is received in the reply to the first request. It is the first response with the wrong header or the second response instead of the first one? what happens to the second response ? Your client receives it or not ? H2 may be quite confusing because the requests are multiplexed. So you must be careful to not mix the responses.

You said 16ms for your client is very fast for a retry ? what does it means ? your client sends a first request and because of an error it retries it with a new UUID ? Is there any chance you handle the response of the second request thinking it is the one of the first ? You also said it is strange to have both requests sent on the same TCP stream. You mean, on the server side or the client side ?

Finally, it may be good to share your configuration, to be sure there is nothing strange. And, if you have a network capture, could you share it privately ?

**Comment by wtarreau** [comment]
I can imagine a scenario which could trigger this in case we'd have one or several bugs, none of which I'm currently aware of. But let's imagine this:
- client 1 sends a request with UUID1 which is passed to varnish, responding with the same UUID
- for whatever reason (timeout?) haproxy aborts this request to varnish
- for another reason (bug?) the connection to varnish is not destroyed but kept alive
- varnish finally responds on this connection
- client 2 sends a request with UUID2 to the same varnish and reuses the connection, then gets the pending response to UUID1
- client is not happy with this response and aborts (hence CL)

I know it requires many "if" but we have to start somewhere :-)  It also doesn't explain the RST from varnish, which might also be part of the equation to trigger a new request. Maybe it returns a valid response plus a reset and we deliver the response AND retry, thus leaving a secon pending response on the connection for example.

If the problem persists with retry-on disabled, another good test to try to bisect this would be to add "http-reuse never" to the configuration. This will prevent from sharing connections between multiple clients.

**Comment by capflam** [comment]
@wtarreau, the first response was aborted in the data phase. The termination state is `SD`. So there is still the possibility to have the wrong termination state. But not so obvious.

@peterfroehlich, after rereading your last comment, I understand the mixup is on the responses headers this time. So, it is not exactly the same issue you firstly described. It was about a mixup on requests headers. Do you know if it is still a headers mix or a complete responses mix this time ? On the client side, is there any response to the second request ?

**Comment by wtarreau** [comment]
I'm thinking about something, if the client is using H2, there is an interesting difference between H1 an H2 in case the client sends two requests, which is that in H1 they're necessarily in-order while in H2 they are out-of-order. So it is also technically possible that the client itself got confused (or imprecise) about what happens. I mean, let's say that request 2 is replied to before request 1. Maybe the client expects in-order responses and fails when it sees request 2 being replied to before request 1.

@peterfroehlich since you seem to have some traces, you should *really* contact us offline and share what you managed to collect, because we're all trying to guess what you're observing and speculating a lot on what may happen, but at least with some traces and configs we could stop speculating. To be honest, this issue might be the one with the least information in all open issues and because of this it's been kept opened for two months, I'm almost tempted to close it if we cannot make any progress forward :-(

**Comment by peterfroehlich** [comment]
Sorry for the slow communication, but it's really slow going. We are more or less bound on user feedback as it seems the error rate itself isn't terribly conclusive. 
We are running on "retry-on none" with HTX since Thursday and I haven't got a negative report. Error rate went up a bit, but this can be explained with more traffic. So thats a good sign. I will reintroduce a more aggressive cache-control header to force the error to persist longer on the client side if its happening to maybe get some reports. Thats really something I don't want to risk but I too am fed up having that ticket on our team board... 

@veggiespam Would you be able to trial your setup with HTX on and "retry-on none"?

@capflam We can't trace the client side due to not having the pfs keys (can't log them out as this would need a lot of changes in our access log management infra). Thats why I can't really estimate the impact on the client any better. 

@wtarreau I thought varnish closed the connection due to "connection: close" being set by "option http-server-close". I can supply traces with some context, I will contact you on your private email (listed on your homepage).

**Comment by veggiespam** [comment]
At some point this week, we will enable:
```
option http-use-htx
retry-on none
```
on HAProxy 2.0.10.

**Comment by veggiespam** [comment]
Deployed to single node and watching for canary trigger.  Will update as needed.

**Comment by capflam** [comment]
@peterfroehlich, Thanks for your network capture. It was really helpful to diagnose the bug. I pushed a fix in all versions with the HTX. If possible, give it a try. It is a good candidate to fix you problem.

@veggiespam, I don't know if your issue is the same. It seems to be different. But give it a try too. Who knows :)

**Comment by peterfroehlich** [comment]
@capflam Oh wow, thats awesome :) 
I'm not in the office tomorrow and I don't want to stick the colleges to watch over a trunk-build edge cluster, but I will try that out next week. 
I'm really pleased something came out of all that debugging, even if its not the bug we are looking for. ^_^

After reading your commit I am convinced its our problem. Thanks a ton!

**Comment by veggiespam** [comment]
Since enabling htx+retry-on-none, we have had no swaps detected by our trigger and no end user has reported an issue (thus our trigger missed it).  However, this does not mean it is fixed for us since the number of swaps was sparse anyhow.  

I certainly hope the fix from using @peterfroehlich's network capture will cure us also.  I will deploy it - but my client cannot deploy a patch to production; they need a full release.  @capflam, can you please push out 2.0.11 and I will deploy on Monday.  We will leave htx+retry-on for a single node after the upgrade.  

I really hope this does it!

**Comment by veggiespam** [comment]
We discovered that using `option http-use-htx  +  retry-on none` with version 2.0.10 still causes header swaps and I told my client to disable HTX again.  I can see multiple IP addresses crossing the trigger_IDs between each other.  All of the swaps occurred on the single node where we enabled HTX.

I'm still combing through the mess, but it looks like a bunch of remote IPs sent or received the wrong data in a period of 2 minutes.  This occurred  a second time later in the day, but over a period of 1 minute.

My client will try again when version 2.0.11 is released.

**Comment by wtarreau** [comment]
Thank you for your tests. Thanks to Peter's trace, Christopher's fix that
was recently merged definitely addresses such a case where two distinct
requests may occasionally be sent over a connection to a server and the
last one will be assigned the response, resulting in this mixup. We'll
issue 2.0.11 very soon (probably monday or tuesday), we are just checking
everything again in it to make sure we didn't merge any regression since
some fixes are very tricky!

**Comment by veggiespam** [comment]
Thanks for the new version.  We will deploy version 2.0.11 Monday and give you feedback as it comes in.  Thanks.

**Comment by peterfroehlich** [comment]
I can confirm that the bug still showed up after "retry-on none" was activated - the rate was much lower though. 
Edge lb on 2.0.11 are active for the last 24h, no errors till now. I will raise the Cache-Control header settings to force the error to stick again.

**Comment by peterfroehlich** [comment]
We are running 2.0.11 with HTX on all LBs now, no errors at all. I consider this fixed. 
Thank you all! ❤️

**Comment by capflam** [comment]
@peterfroehlich, thanks for confirming. We will wait for @veggiespam feedback to be sure the bug is fixed.

**Comment by veggiespam** [comment]
The system was unintentionally configured with the default of `retry-on conn-failure` by mistake instead of `none` as Peter did now and we did in the past.  In a way, this is good as we are testing different configurations than Peter.  So far, no header swaps.

A concern I have is from @peterfroehlich's in https://github.com/haproxy/haproxy/issues/290#issuecomment-566447572  - you had the issue and then it disappeared with a different Cache header?  As my client has done nothing in Caching, what is the impact for our system?  If you had one swap and then no swaps after a caching change, then this is not truly fixed.  Or - was you swap a false positive in that instance.

Thanks for helping me to understand - let's not close this ticket just yet.

**Comment by veggiespam** [comment]
No triggered header mix-ups since Monday.  But as the holidays usually have decreased traffic, this less likely to appear up anyhow.  I'm leaning towards "fixed" but want to give this more burn-in time.

Due to the holiday staffing issues, my client will keep HTX only on the one canary node.  If no problems are found in the next few weeks, then I will have them migrate to all nodes on Jan 9 or 10.

Assuming this is fixed @capflam, can you publish a CVE as there were multiple instances of data leakage and account take overs described by many of us across a few github tickets.

**Comment by wtarreau** [comment]
OK thanks for the feedback. I'm switching it to feedback so that it doesn't remain high on the todo list.

**Comment by hgshoggins** [comment]
I hope I'm not hijacking this issue, but I'm experiencing the same problem without HTX enabled (or at least not explicitly). My backend servers are only serving `http/1.1`, but I'm offering `h2` at my frontends. I really thought that was my backend servers that had gone crazy, but the symptoms described here and in #292 are exactly the same: sometimes our served web pages are jammed by binary PNG, JPEG or else served instead of `text/html` content.

I'm using `haproxy-2.0.10` from `Fedora 31` distribution, conclusion comment in #292 seemed to say that the issue was fixed with `2.0.7` but it's still happening to me.

My `bind` instructions for my `frontend` sections have the `alpn h2,http/1.1` options. I just removed it to remain with `http/1.1 TLS` only, so I guess I can temporarily overcome this issue with this, but I'm not completely sure I understood all the parameters of the issue, so not sure if I'm doing it right.

**Comment by Feder1co5oave** [comment]
@hgshoggins please try 2.0.11 first

**Comment by veggiespam** [comment]
@hgshoggins - HTX is enabled by default in the 2.0.x series.  The general belief is that 2.0.11 may have corrected the issue (many of us are doing burn ins to validate).  

So, either upgrade to 2.0.11 per @Feder1co5oave's suggestion or, if you are unable to upgrade just yet, set `no option http-use-htx` in your current 2.0.10.

**Comment by hgshoggins** [comment]
Thank you, I came with the latter solution, waiting for a update from upstream in my distro.

**Comment by veggiespam** [comment]
My client now uses HTX on all endpoints with version 2.0.11 and has seen no header swaps.  We are calling this problem fixed.

I confidentially spoke with friends in the finance industry and they have run into header swaps involving `set-cookie`.  They had users sending screen shots of bank account information of other users to the bank's customer support teams.  This bug is very hard to replicate and to detect, so the likelihood vector is low.  But the impact of account take-over is severe (e.g.; wire transfer money out).  You need to issue a security alert about this problem for version 1.9/2.0/2.1 to upgrade to the minimum appropriate patch level.  Calling this a security problem instead of "BUG/MEDIUM" will push people to upgrade.  Doing this will also light a fire under Linux distros to update more quickly.

Thank you again for fixing the issue.

**Comment by wtarreau** [comment]
Hi,

first, thanks for your detailed feedback.

Just to clarify regarding what your friends observed, there's no "headers swap" but "response swaps" due to this bug, but the swap necessarily happens within the same client connection because the server-side connection is attached to the client's session. It may happen that your friends have stacked multiple haproxy layers however and a failure on the last one could indeed trigger the problem. Otherwise maybe they were facing another bug that was since fixed.

Now regarding assigning a CVE or not. Feel free to try to get one and handle it if you wish, personally I finally gave up on this bureaucratic circus for multiple reasons ranging from the randomness of the responses to the impossibility to get their wrong and misleading rewording fixed even after it confuses everyone with false information. Distros usually try to follow bugs more seriously, or you'll know which ones not to use once you figure it's their policy not to apply fixes (at least distros shipping haproxy 2.0 packages are up to date).

In addition if you know people who still use CVEs as a signal to upgrade, you should teach them that while it was a passably acceptable practice in the 1990s when software did not evolve much and security issues were the only bugs one could meet, it's totally obsolete in 2020 when identified security issues are just marginal while technical and functional bugs are 99%, some of which may or may not have a security impact with nobody knowing, and all of them having a real business impact. And those who continue with this bad practice after having being explained the above must be fired.

CVEs in 2020 are only an alternative for quality used in businesses driven by incompetent managers.
